### PR TITLE
Compute the correct layout for variants of uninhabited enums

### LINKED ISSUE #69191: beta regression: ICE on Tried to access field 0 of union Layout
- [`rotor v0.6.3`](https://crater-reports.s3.amazonaws.com/beta-1.42-1/beta-2020-02-05/reg/rotor-0.6.3/log.txt)
- [`rotor-dns v0.4.0`](https://crater-reports.s3.amazonaws.com/beta-1.42-1/beta-2020-02-05/reg/rotor-dns-0.4.0/log.txt)
- [`rotor-tools v0.3.2`](https://crater-reports.s3.amazonaws.com/beta-1.42-1/beta-2020-02-05/reg/rotor-tools-0.3.2/log.txt)

cc @oli-obk 

I suspect (but have not verified) that this would be fixed by a backport of https://github.com/rust-lang/rust/pull/64987.

### LINKED ISSUE #69763: Layout and MIR field accesses are incoherent
I propose that we should have the following invariant: Whenever, anywhere in MIR, there is a [field projection](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/mir/enum.ProjectionElem.html#variant.Field) with some index `i`, then `i` is actually a valid field index in the (statically known) layout of the type being projected into.

Right now, [this is not the case](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=6d0730b4f8fd170d74f5d2eb131e5e04). This code:
```rust
pub enum Void {}

enum UninhabitedUnivariant { _Variant(Void), }

fn main() {
    let _seed: UninhabitedUnivariant = None.unwrap();
    match _seed {
        UninhabitedUnivariant::_Variant(_x) => {}
    }
}
```
generates the following MIR:
```
    bb1: {
        StorageDead(_2);                 // bb1[0]: scope 0 at src/main.rs:6:52: 6:53
        StorageLive(_3);                 // bb1[1]: scope 1 at src/main.rs:8:41: 8:43
        _3 = move ((_1 as _Variant).0: Void); // bb1[2]: scope 1 at src/main.rs:8:41: 8:43
        StorageDead(_3);                 // bb1[3]: scope 1 at src/main.rs:9:5: 9:6
        StorageDead(_1);                 // bb1[4]: scope 0 at src/main.rs:10:1: 10:2
        return;                          // bb1[5]: scope 0 at src/main.rs:10:2: 10:2
    }
```
but the layout of `_Variant` is that of a union with 0 fields.

A consequence of this incoherence is that all codegen backends and Miri (and possibly more MIR consumers) all need to be on their guard when considering field projections, always protecting against the case where the projection is ill-formed (usually that happens by special-casing uninhabited types before considering the projection).

I think instead of putting that burden on every MIR consumer, we should fix either the layout or the MIR to not violate the invariant in the first place.

Cc @eddyb with whom I anyway recently talked about those odd `Union` layouts. Is there some fundamental issue with the invariant I am proposing? Also Cc @oli-obk @wesleywiser 

If we decide to go with this, we should
* [x] Revert https://github.com/rust-lang/rust/pull/69753
* [ ] Check for other cases in Miri/const-prop that just work around such broken MIR/layout
* [x] Check for the corresponding cases in codegen and other MIR consumers -- mostly, add assertion [here](https://github.com/rust-lang/rust/blob/865b44a3e330f3ef8be0f6edf69896c9ed957ac0/src/librustc_target/abi/mod.rs#L663).

### PR DESCRIPTION
**Pull request by oli-obk**
r? @eddyb
cc @RalfJung

fixes #69191
cc #69763

### CODE REVIEW

**Review by eddyb — APPROVED** [review]

File: src/librustc/ty/layout.rs [hunk 1]
```diff
@@ -2023,7 +2023,14 @@ where
 {
     fn for_variant(this: TyLayout<'tcx>, cx: &C, variant_index: VariantIdx) -> TyLayout<'tcx> {
         let details = match this.variants {
-            Variants::Single { index } if index == variant_index => this.details,
+            Variants::Single { index }
+                // If all variants but one are uninhabited, the variant layout is the enum layout.
+                if index == variant_index &&
+                // Don't confuse variants of uninhabited enums with the enum itself.
```
Comment (RalfJung) [inline src/librustc/ty/layout.rs]: Maybe reference https://github.com/rust-lang/rust/issues/69763 to explain why that's a problem

**Comment by RalfJung** [comment]
> fixes #69763

That issue also says to check other code for unnecessary `is_uninhabited` checks, so maybe we shouldn't close it yet?
@eddyb seems to disagree that we should get rid of these checks in codegen, but I want to at least audit Miri for this.

**Comment by eddyb** [comment]
Oh yeah I missed the bit about #69763, it shouldn't be closed until the plan at the end of https://github.com/rust-lang/rust/issues/69763#issuecomment-595699745, or something similarly thorough, is implemented.

File: src/librustc_mir/interpret/operand.rs [hunk 2]
```diff
@@ -361,7 +361,7 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
     ) -> InterpResult<'tcx, OpTy<'tcx, M::PointerTag>> {
         let base = match op.try_as_mplace(self) {
             Ok(mplace) => {
-                // The easy case
+                // We can reuse the mplace field compuation logic for indirect operands
```
Comment (lqd) [inline src/librustc_mir/interpret/operand.rs]:
```suggestion
                // We can reuse the mplace field computation logic for indirect operands
```

File: src/test/ui/consts/issue-69191.rs [hunk 3]
```diff
@@ -36,6 +36,12 @@ fn main() {
         UninhabitedUnivariant2::_Warriont(_x) => {}
     }
 
+    let _seed: UninhabitedUnivariant2 = None.unwrap();
+    match _seed {
+        UninhabitedUnivariant2::_Warriont(_x) => {}
```
Comment (RalfJung) [inline src/test/ui/consts/issue-69191.rs]: Well while you are at it, what about just the first on its own?^^

**Comment by nikomatsakis** [comment]
Discussed in today's meeting and decided against backporting this, since #69753 fixes the same problem in a narrow way (though this change is still desired).

**Comment by RalfJung** [comment]
@bors r- because https://github.com/rust-lang/rust/pull/69753 goes first, and this PR should be rebased on top of that and revert the Miri changes.

Also @oli-obk could you add a corresponding assertion [here](https://github.com/rust-lang/rust/blob/865b44a3e330f3ef8be0f6edf69896c9ed957ac0/src/librustc_target/abi/mod.rs#L663) that checks if `i` is in-bounds for the fields of this union? I think that should make sure codegen also notices these kinds of MIR incoherences.

**Comment by eddyb** [comment]
> Also @oli-obk could you add a corresponding assertion here that checks if i is in-bounds for the fields of this union? I think that should make sure codegen also notices these kinds of MIR incoherences.

cc https://github.com/rust-lang/rust/pull/64987#discussion_r330986392

**Comment by RalfJung** [comment]
@eddyb that PR actually [did add the assertion](https://github.com/rust-lang/rust/pull/64987/files#diff-1240c4024b5a96328bb0436c2344a020R742), but it seems someone removed it again later?

**Comment by RalfJung** [comment]
Ah it got reverted in https://github.com/rust-lang/rust/pull/66250

File: src/librustc_mir/interpret/operand.rs [hunk 4]
```diff
@@ -356,7 +356,7 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
     ) -> InterpResult<'tcx, OpTy<'tcx, M::PointerTag>> {
         let base = match op.try_as_mplace(self) {
             Ok(mplace) => {
-                // The easy case
+                // We can reuse the mplace field computation logic for indirect operands
```
Comment (RalfJung) [inline src/librustc_mir/interpret/operand.rs]:
```suggestion
                // We can reuse the mplace field computation logic for indirect operands.
```

**Comment by RalfJung** [comment]
LGTM, but I cannot really judge the changes in `layout.rs`.

File: src/librustc_target/abi/mod.rs [hunk 5]
```diff
@@ -660,7 +660,11 @@ impl FieldPlacement {
 
     pub fn offset(&self, i: usize) -> Size {
         match *self {
-            FieldPlacement::Union(_) => Size::ZERO,
+            FieldPlacement::Union(count) => {
+                assert!(i < count,
+                        "Tried to access field {} of union with {} fields", i, count);
```
Comment (eddyb) [inline src/librustc_target/abi/mod.rs]: We should eventually figure out an uniform style across the compiler for this sort of thing.

I'm personally for `// Foo bar.` for comments but `bug!("foo bar");` for ICEs.

(but you can leave this as-is)

**Comment by oli-obk** [comment]
`layout.rs` was already signed off earlier in this PR

@bors r=eddyb,RalfJung

**Review by eddyb — APPROVED** [review]

**Pull request merged**